### ISSUE TITLE
Make stringify output a well-formed Unicode string (matching JSON.stringify)

### ISSUE THREAD

**Issue by mathiasbynens** [fastify/fast-json-stringify#147]
We recently changed `JSON.stringify` to output well-formed strings, even when the input contains lone surrogates: https://github.com/tc39/proposal-well-formed-stringify

Example:

```js
const fastJson = require('fast-json-stringify')
const stringify = fastJson({
  title: 'Example Schema',
  type: 'object',
  properties: {
    firstName: {
      type: 'string'
    },
    lastName: {
      type: 'string'
    },
    age: {
      description: 'Age in years',
      type: 'integer'
    },
    reg: {
      type: 'string'
    }
  }
})

console.log(stringify({
  firstName: 'Matteo\uD800', // <-- lone surrogate at the end
  lastName: 'Collina',
  age: 32,
  reg: /"([^"]|\\")*"/
}))
```

Currently, the surrogate appears unescaped in the output.

**Comment by mcollina** [comment]
Thanks for opening this up. I think we can handle this by changing this [ternary](https://github.com/fastify/fast-json-stringify/blob/d1677f73b5825343ad1a4a0b6652bfb014a4a4e6/index.js#L261) so that we bail to `JSON.stringify()` in those cases.

**Comment by mathiasbynens** [comment]
The annoying thing is that you need to check the previous/next character as well to figure out if it's indeed a lone surrogate, and not a surrogate pair. Here's how we implemented in in V8, FWIW: https://github.com/v8/v8/commit/a11e05be3ac91723becfbb26344f955b333d09e0

**Comment by srmarjani** [comment]
Hi 
What is difference between  lone surrogate and surrogate pair?

**Comment by mcollina** [comment]
An unicode character is composed of more than 1 byte, starting with a prefix. I think a lone surrogate is when you have the prefix byte without the follow-up.

**Comment by srmarjani** [comment]
OK.
My knowledge about unicode is low.
I must do some research before doing some code on this issue.

**Comment by mathiasbynens** [comment]
Copying something I wrote on <https://mathiasbynens.be/notes/javascript-encoding#surrogate-pairs> in case it provides more background:

> Characters outside the BMP, e.g. _U+1D306 tetragram for centre_ (`𝌆`), can only be encoded in UTF-16 using two 16-bit code units: `0xD834 0xDF06`. This is called a _surrogate pair_. Note that a surrogate pair only represents a single character.
> 
> The first code unit of a surrogate pair is always in the range from `0xD800` to `0xDBFF`, and is called a _high surrogate_ or a _lead surrogate_.
> 
> The second code unit of a surrogate pair is always in the range from `0xDC00` to `0xDFFF`, and is called a _low surrogate_ or a _trail surrogate_.
> 
> UCS-2 lacks the concept of surrogate pairs, and therefore interprets `0xD834 0xDF06` (the previous UTF-16 encoding) as two separate characters.

A lone surrogate is a surrogate that occurs by itself, i.e. it is not part of a pair.

**Comment by srmarjani** [comment]
Hi @mcollina 

I don't get this code. Where does 42 come from?

` if (str.length < 42) {
    return $asStringSmall(str)
  } else {
    return JSON.stringify(str)
  }`

I read the code that @mathiasbynens  mentioned for V8.
As I understood in V8, string examined character by character and they have in mind unicode.
But in fast-json-stringify I think you've considered only ascii code. Am I wrong?

**Comment by mcollina** [comment]
Short strings are faster to escape manually rather than calling `JSON.stringify()`. I had to pick a number, and I picked 42. The main reason we have this is because of object keys: those are short and using this hack increase the throughput significantly.

Note that there is a fallback logic to go to `JSON.stringify()`:
https://github.com/fastify/fast-json-stringify/blob/master/index.js#L261. We should change that logic to defer to `JSON.stringify()` if we detect a surrogate.
